python_ta/debug/recursion_table.py

```diff
@@ -0,0 +1,161 @@
+"""
+Table data structure that prints a nicely formatted table
+for a recursive function
+"""
+from __future__ import annotations
+
+import inspect
+import sys
+import types
+from typing import Any, Callable
+
+import tabulate
+
+from python_ta.util.tree import Tree
+
+
+def clean_frame_variables(frame: types.FrameType) -> dict[str, Any]:
+    """remove the local variables from the frame's locals and keep only the
+    parameters
+    """
+    raw_variables = frame.f_locals
+    parameters = inspect.getargvalues(frame).args
+    # not mutating the local variables to avoid unintended effects
+    cleaned_variables = {param: raw_variables[param] for param in parameters}
+    return cleaned_variables
+
+
+class RecursionTable:
+    """
+    Class used as a form of print debugging to analyze the inputs
+    and return values for a recursive function
+
+    Instance attributes:
+        frames_mapping: a mapping between the frame for a
+            recursive function and their return values and inputs
+        frames_ordered: a list to conserve the order in which the
+            frames were created
+        function_name: name of the function being traced
+        _trees: mapping of the frames to the corresponding tree
+            representing the function call
+    """
+
+    frames_mapping: dict[types.FrameType, list]
```

Comment:
First, change the name of the attribute to `frame_data`.

More involved change: Instead of the dict values being lists, have them be dictionaries that store the individual pieces of data in the same structure as the eventual returned table, plus an extra field for the computed string representation of the call.

More concretely, the type annotation should be `dict[types.FrameType, dict[str, Any]]`. For the `fact` example each associated value should look like:

```python
{
  'n': ...,  # e.g., 4
  'return value': ...,  # e.g., '24'
  'called by': ...,  # e.g., 'fact(5)'
  'call string': ...  # e.g., 'fact(4)'
}
```

That is, move some of the logic in the `get_recursive_dict` method into the `_record_call` method to provide more structure to the data at the time when the call is initially recorded.

Finally, note that I've asked you to store `'call string'`; if you do so, you should be able to *look up previously-computed call strings* rather than recomputing them for every "parent" call string.